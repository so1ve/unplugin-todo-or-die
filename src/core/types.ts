import type { FilterPattern } from "@rollup/pluginutils";

export interface Options {
	include?: FilterPattern;
	exclude?: FilterPattern;
}
