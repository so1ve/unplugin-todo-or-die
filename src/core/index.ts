import type { UnpluginFactory } from "unplugin";
import { createUnplugin } from "unplugin";

import type { Options } from "./types";

export const unpluginFactory: UnpluginFactory<Options | undefined> = (
	_options,
) => ({
	name: "pkg-name",
	transform(code) {
		return code;
	},
});

export const unplugin = /* #__PURE__ */ createUnplugin(unpluginFactory);

export default unplugin;
