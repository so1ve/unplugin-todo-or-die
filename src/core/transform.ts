import MagicString from "magic-string";


// eslint-disable-next-line regexp/no-super-linear-backtracking
const todoReg = /^\s*//\s*TODO::expires?\((\d{4}-\d{2}-\d{2})\):\s*(.*)$/g;

interface Todo {
	expires: number;
	content: string;
	start:number
	end: number
}

function parseComment(match: RegExpExecArray): Todo | undefined {
	
	const [comment, date, content] = match;
	if ( date) {
		const expires = new Date(date).getTime();

		return {
			expires,
			content,
			start:match.index,end:match.index+comment.length
		};
	}
}

const generateDieCode = (todo: Todo) =>
	`if (Date.now() > ${todo.expires}) throw new Error("TODO expired: ${todo.content}");`;



export function transform(code: string, id: string) {
	const s = new MagicString(code);
		const matches=code.matchAll(todoReg)
	const todos = matches.map(parseComment).filter(Boolean) as Todo[];

	for (const todo of todos) {
		s.overwrite(todo.start, todo.end, generateDieCode(todo));
	}

	return {
		code: s.toString(),
		map: s.generateMap({ hires: true }),
	};
}
