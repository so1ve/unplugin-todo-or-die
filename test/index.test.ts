import { describe, expect, it } from "vitest";

import { transform } from "../src/core/transform";

const testTransform = (code: string) => transform(code, "foo.ts");

describe("should", () => {
	it("transform", () => {
		expect(
			testTransform(`
		// TODO::expires(2000-01-01): do something
		`),
		).toMatchSnapshot();
	});

	it("transform another case", () => {
		expect(
			testTransform(`
		// TODO::expire(2000-01-01): do something
		`),
		).toMatchSnapshot();
	});
});
