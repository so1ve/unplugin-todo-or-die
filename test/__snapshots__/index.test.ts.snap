// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should > transform 1`] = `
{
  "code": "
		if (Date.now() > 946684800000) throw new Error(\\"TODO expired: do something\\");
		",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA,CAAC,CAAC,6EAA0C;AAC5C,CAAC",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > transform another case 1`] = `
{
  "code": "
		if (Date.now() > 946684800000) throw new Error(\\"TODO expired: do something\\");
		",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA,CAAC,CAAC,6EAAyC;AAC3C,CAAC",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
